packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js

```diff
@@ -104,7 +104,10 @@ export default function useAutocomplete(props) {
 
   function setHighlightedIndex(index, mouse = false) {
     highlightedIndexRef.current = index;
-    inputRef.current.setAttribute('aria-activedescendant', `${id}-option-${index}`);
+    // does the index exist?
+    if (index !== -1) {
+      inputRef.current.setAttribute('aria-activedescendant', `${id}-option-${index}`);
```

Comment:
I would propose the following alternative solution:

```diff
diff --git a/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js b/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js
index 31bee7fef..83b71605a 100644
--- a/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js
+++ b/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js
@@ -104,7 +104,11 @@ export default function useAutocomplete(props) {

   function setHighlightedIndex(index, mouse = false) {
     highlightedIndexRef.current = index;
-    inputRef.current.setAttribute('aria-activedescendant', `${id}-option-${index}`);
+    if (index === -1) {
+      inputRef.current.removeAttribute('aria-activedescendant');
+    } else {
+      inputRef.current.setAttribute('aria-activedescendant', `${id}-option-${index}`);
+    }

     if (!listboxRef.current) {
       return;
```

This approach should help to solve two edge cases:

- the `includeInputinList` prop allows to include the input in the focus ring. I think that we should keep the attribute in sync.
- the `useAutocomplete` hook allows keeping the listbox visible when the input is blurred. I'm wondering if we should keep the aria-activedescendant on the right index (even if not in the DOM)?

> If the design calls for a specific element to be focused the next time a user moves focus into the composite with Tab or Shift+Tab, check if aria-activedescendant is referring to that target element when the container loses focus. If it is not, set aria-activedescendant to refer to the target element.

https://www.w3.org/TR/wai-aria-practices/#kbd_focus_activedescendant

Reply:
If you have a test, sure. Otherwise I would leave it as is and remove this approach in the future. Mixing react with imperative attribute manipulation defeats the purpose of using react in the first place.

Reply:
@eps1lon Did you delete the comment? Do you still want me to contribute a test case on this branch for the `includeInputinList` case?

Reply:
Huh weird. Yeah please add a test if you already know what this should be doing. Intuitively `includeInputinList` sounds pretty confusing since textbox shouldn't be considered focused with the listbox. Sounds like it's mixing semantics with styling and the styling being misleading.

Reply:
This prop is meant to accommodate two practices I have seen with the `freeSolor` use case (I hope common enough). Google Chrome URL bar and [Google Search bar](https://www.google.com/) are two good examples. The first one would use `includeInputInList={false}` while the second would have `includeInputInList={true}`.

I can take two other examples [react-select](https://react-select.com/) and [reach-ui](https://reacttraining.com/reach-ui/combobox). The first one would use `includeInputInList={false}` while the second would have `includeInputInList={true}`.

Here is the test case I have in mind for concern #1:

```jsx
describe('includeInputInList', () => {
  it('can move the focus back to the input', () => {
    const { getAllByRole, getByRole } = render(
      <Autocomplete
        includeInputInList
        options={['one', 'two  ']}
        renderInput={params => <TextField autoFocus {...params} />}
      />,
    );

    fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
    const textbox = getByRole('textbox');
    const options = getAllByRole('option');
    expect(textbox).to.have.attribute('aria-activedescendant', options[0].getAttribute('id'));
    fireEvent.keyDown(document.activeElement, { key: 'ArrowUp' });
    expect(textbox).not.to.have.attribute('aria-activedescendant');
  });
});
```

Reply:
> This prop is meant to accommodate two practices I have seen (I hope common enough)

You didn't describe what this practice is. You blindly copied it as far as I can tell.

As far as I can tell this means I can move out of the listbox with arrow keys which I'd need a rationale not just c&p. I suspect being able to move focus out of a list with arrow keys is frustrating since it's never used.

Either way add the test in a PR so that I can rebase the work and then we can discuss if there's anything to this prop besides bloating API surface.

Reply:
The practice is to return the focus to the textbox vs not returning it. The rationale is two-fold:
1. I think that both patterns are common enough to justify implementing them, we leave the rationale for using one vs another to the users. However, we need a default value for the `includeInputInList` boolean. We can solve it using our "taste", that hopefully, our users will correct if we pick wrong. So one approach could be to rely on the **collective intelligence**. This is the rationale I personally prefer, I think that it should be our ultimate goal but it can be challenging to follow.
2. Another rationale that you might like better is to refer to WAI-ARIA:

> Up Arrow: Moves focus to and selects the previous option. If focus is on the first option, either returns focus to the textbox or does nothing.

In practice, the WAI-ARIA example implementations don't follow any of the 2 possible options, it wraps. The Autocomplete implements the 3rd undocumented option by default, but offer the "focus the textbox" with `includeInputInList` and the "does nothing" with `disableListWrap`. Both are opt-in.

> Either way add the test in a PR so that I can rebase the work and then we can discuss if there's anything to this prop besides bloating API surface.

What do you think of removing the source changes from this pull request, to focus on the addition of the new tests? Then, I can open a pull request to add this new failing test case and update the source to fix it?

Reply:
The source changes isn't just some random addition because I was bored. I don't change source without testing it.

Reply:
Merging this. We can restore the behavior in a follow-up PR with a test.